### PR TITLE
Raise error when observation keys and observation space keys don't match

### LINKED ISSUE #944: [Bug] in `env_checker.py`
### 🐛 Bug

When the `observation_space` is a `Dict`, only the keys in the `observation_space.spaces.keys()` are checked with the keys returned in the observation. If there is a key in the returned observation, which is not present in the `observation_space.spaces.keys()` then no error is thrown, which should not be the case ideally.

Code snippets from `env_checker.py` for your reference

```python
# for reset
if isinstance(observation_space, spaces.Dict):
        assert isinstance(obs, dict), "The observation returned by `reset()` must be a dictionary"
        for key in observation_space.spaces.keys():
            try:
                _check_obs(obs[key], observation_space.spaces[key], "reset")
            except AssertionError as e:
                raise AssertionError(f"Error while checking key={key}: " + str(e)) from e
``` 

```python
# for step
if isinstance(observation_space, spaces.Dict):
        assert isinstance(obs, dict), "The observation returned by `step()` must be a dictionary"
        for key in observation_space.spaces.keys():
            try:
                _check_obs(obs[key], observation_space.spaces[key], "step")
            except AssertionError as e:
                raise AssertionError(f"Error while checking key={key}: " + str(e)) from e
```

I'll work on fixing it, but wanted your approval before that

### PR DESCRIPTION
**Pull request by qgallouedec**
## Description

See #944. Duplicate #947, but simpler implementation.

closes #944 
closes #947 

## Motivation and Context

- [x] I have raised an issue to propose this change ([required](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) for new features and bug fixes)

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)

## Checklist:

- [x] I've read the [CONTRIBUTION](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) guide (**required**)
- [ ] I have updated the changelog accordingly (**required**).
- [ ] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.
- [x] I have reformatted the code using `make format` (**required**)
- [ ] I have checked the codestyle using `make check-codestyle` and `make lint` (**required**)
- [ ] I have ensured `make pytest` and `make type` both pass. (**required**)
- [ ] I have checked that the documentation builds using `make doc` (**required**)

Note: You can run most of the checks using `make commit-checks`.

Note: we are using a maximum length of 127 characters per line

### CODE REVIEW

File: stable_baselines3/common/env_checker.py [hunk 1]
```diff
@@ -143,6 +143,10 @@ def _check_returned_values(env: gym.Env, observation_space: spaces.Space, action
 
     if isinstance(observation_space, spaces.Dict):
         assert isinstance(obs, dict), "The observation returned by `reset()` must be a dictionary"
+
+        if not obs.keys() == observation_space.spaces.keys():
+            raise AssertionError(f"The observation keys returned by `reset()` must match the observation space keys")
```
Comment (araffin) [inline stable_baselines3/common/env_checker.py:148]: you should also display the diffrence (current the f-string is not used)

File: stable_baselines3/common/env_checker.py [hunk 2]
```diff
@@ -162,6 +166,10 @@ def _check_returned_values(env: gym.Env, observation_space: spaces.Space, action
 
     if isinstance(observation_space, spaces.Dict):
         assert isinstance(obs, dict), "The observation returned by `step()` must be a dictionary"
+
+        if not obs.keys() == observation_space.spaces.keys():
+            raise AssertionError(f"The observation keys returned by `step()` must match the observation space keys")
```
Comment (araffin) [inline stable_baselines3/common/env_checker.py:171]: idem ;)

File: stable_baselines3/common/env_checker.py (hunk 1 above)
Comment (qgallouedec) [reply stable_baselines3/common/env_checker.py:148]: Something like 

    Observation space contains key 'foo', but not the observation returned by step()

?

Comment (araffin) [reply stable_baselines3/common/env_checker.py:148]: or simply: `f"{obs.keys()} != {observation_space.spaces.keys()}"`

**Comment by qgallouedec** [comment]
| Method | Key "img" missing from | Output before | Output now |
|-|-|-|-|
| Reset | observation | `KeyError: 'img'` | ``AssertionError: The observation keys returned by `reset()` must match the observation space keys: dict_keys(['img', 'vec']) != odict_keys(['vec'])`` |
| Reset | observation space | No error | ``AssertionError: The observation keys returned by `reset()` must match the observation space keys: odict_keys(['vec']) != odict_keys(['img', 'vec'])`` |
| Step   | observation | `KeyError: 'img'` | ``AssertionError: The observation keys returned by `step()` must match the observation space keys: dict_keys(['img', 'vec']) != odict_keys(['vec'])`` |
| Step   | observation space | No error | ``AssertionError: The observation keys returned by `step()` must match the observation space keys: odict_keys(['vec']) != odict_keys(['img', 'vec'])`` |

## To reproduce

```python
from stable_baselines3.common.envs import SimpleMultiObsEnv
from stable_baselines3.common.env_checker import check_env

env = SimpleMultiObsEnv()

wrong_obs = env.observation_space.sample()
wrong_obs.pop("img") # remove a key so that the obs becomes wrong

def wrong_reset():
        return wrong_obs

env.reset = wrong_reset
check_env(env)
```

**Review by araffin — APPROVED** [review]
LGTM, thanks =)
please don't forget to update the changelog ;)

**Pull request merged**